database/src/ops/block.rs

```diff
@@ -1,89 +1,492 @@
+                } else {
+                    add_output(
+                        amount,
+                        &Output {
+                            key,
+                            height,
+                            output_flags,
+                            tx_idx,
+                        },
+                        tables,
+                    )?;
+                }
+            // RingCT outputs.
+            } else {
+                let commitment = tx.rct_signatures.base.commitments[i].compress().to_bytes();
+                add_rct_output(
+                    &RctOutput {
+                        key,
+                        height,
+                        output_flags,
+                        tx_idx,
+                        commitment,
+                    },
+                    tables.rct_outputs_mut(),
+                )?;
+            }
+        }
+    }
+
+    Ok(())
 }
 
-/// TODO
-pub fn get_block_cumulative_rct_outputs() {
-    todo!()
+//---------------------------------------------------------------------------------------------------- `pop_block_*`
+/// Remove the top/latest block from the database.
+///
+/// The removed block's height and hash are returned.
+#[doc = doc_error!()]
+// no inline, too big
+pub fn pop_block(tables: &mut impl TablesMut) -> Result<(BlockHeight, BlockHash), RuntimeError> {
+    // Remove block data from tables.
+    let (block_height, block_hash) = {
+        let (block_height, block_info) = tables.block_infos_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    };
+
+    // Block heights.
+    tables.block_heights_mut().delete(&block_hash)?;
+
+    // Block blobs.
+    // We deserialize the block blob into a `Block`, such
+    // that we can remove the associated transactions later.
+    let block_blob = tables.block_blobs_mut().take(&block_height)?.0;
+    let block = Block::read(&mut block_blob.as_slice())?;
+
+    // Transaction & Outputs.
+    for tx_hash in block.txs {
+        let (tx_id, tx_blob) = remove_tx(&tx_hash, tables)?;
+        let tx = Transaction::read(&mut tx_blob.0.as_slice())?;
+
+        // Is this a miner transaction?
+        let mut miner_tx = false;
+        for inputs in &tx.prefix.inputs {
+            match inputs {
+                // Key images.
+                Input::ToKey { key_image, .. } => {
+                    let result =
+                        remove_key_image(key_image.compress().as_bytes(), tables.key_images_mut());
+
+                    // TODO: all test tx's have the same key image so the 1st
+                    // removal removes everything - fix me after we have real data.
+                    if cfg!(test) {
+                        match result {
+                            Ok(()) | Err(RuntimeError::KeyNotFound) => (),
+                            Err(e) => return Err(e),
+                        }
+                    } else {
+                        result?;
+                    }
```

Comment:
Will be fixed after we have real test data - the current ones all have the same key image.